datasets/flwr_datasets/partitioner/distribution_partitioner.py

```diff
@@ -0,0 +1,423 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Distribution partitioner class that works with Hugging Face Datasets."""
+
+
+from collections import Counter
+from typing import Dict, List, Optional, Union
+
+import numpy as np
+
+import datasets
+from flwr_datasets.common.typing import NDArray, NDArrayFloat, NDArrayInt
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class DistributionPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on a distribution.
+
+    Inspired from implementations of Li et al. Federated Optimization in
+    Heterogeneous Networks (2020) https://arxiv.org/abs/1812.06127.
+
+    Given a 2-dimensional user-specified distribution, the algorithm splits the dataset
+    for each unique label per partition where each label is assigned to the partitions
+    in a deterministic pathological manner. The 1st dimension is the number of unique
+    labels and the 2nd-dimension is the number of buckets into which the samples
+    associated with each label will be divided. That is, given a distribution array of
+    shape,
+                           `num_unique_labels_per_partition` x `num_partitions`
+    ( `num_unique_labels`, ---------------------------------------------------- ),
+                                          `num_unique_labels`
+    the label_id at the i'th row is assigned to the partition_id based on the following
+    approach.
+
+    First, for an i'th row, generate a list of `id`s according to the formula:
+        id = alpha + beta
+    where,
+        alpha = (i - num_unique_labels_per_partition + 1) \
+                 + (j % num_unique_labels_per_partition),
+        alpha = alpha + (alpha >= 0 ? 0 : num_unique_labels),
+        beta = num_unique_labels * (j // num_unique_labels_per_partition)
+    and j in {0, 1, 2, ..., `num_columns`}. Then, sort the list of `id`s in ascending
+    order. The j'th index in this sorted list corresponds to the partition_id that the
+    i'th unique label (and the underlying distribution array value) will be assigned to.
+    So, for a dataset with 10 unique labels and a configuration with 20 partitions and
+    2 unique labels per partition, the 0'th row of the distribution array (corresponding
+    to class 0) will be assigned to partitions [0, 9, 10, 19], 1st row (class 1) to
+    [0, 1, 10, 11], 2nd row (class 2) to [1, 2, 11, 12], 3rd row (class 3) to
+    [2, 3, 12, 13], etc ... . Alternatively, the distribution can be interpreted as
+    partition 0 having classes 0 and 1, partition 1 having classes 1 and 2, partition 2
+    having classes 2 and 3, etc ... The list representing the unique labels is sorted
+    in ascending order.
+
+    Parameters
+    ----------
+    distribution_array : Union[NDArrayInt, NDArrayFloat]
```

Comment:
I'm thinking we should also allow for this to be a list of lists. We'd convert it into a numpy array in `_check_distribution_array_shape_if_needed()` if it's not one already. What do you think?

Reply:
@jafermarq I like that option. How about we extend it in a following PR? Merging this PR will unblock a few of our activities, and giving a list of lists is a nice enhancement that can follow this PR.